implement boolean parsing for entity files

bmorel opened the pull request:
This reduces "Fix errors and warnings in included entity definition files".

Currently, parseInteger is used to parse boolans, which of course can not work, since booleans are either True or False, and thus text.

I took the freedom to implement 4 variants of true: `True`, `T`, `true`, and `t`, plus numbers (true != 0).
This can likely be improved upon though, i.e. by rejecting invalid text values, but I'd rather wait for feedback and fix all issues in one go.

kduske commented:
Why do we need to support so many variants?

bmorel commented:
> Why do we need to support so many variants?

No real need, but since I was there I though that I could as well add those variants that I've seen used in various places. I could just remove all variants to make something stricter. Your choice.

kduske commented:
Do not force push into existing PRs while they are being reviewed, please.

kduske commented:
> > Why do we need to support so many variants?
> 
> No real need, but since I was there I though that I could as well add those variants that I've seen used in various places. I could just remove all variants to make something stricter. Your choice.

What is "various places"? If you have seen them used in entity definition files, then it's fine. Otherwise, I'd say drop them.

bmorel commented:
> Do not force push into existing PRs while they are being reviewed, please.

Oh, sorry.

I rewrote the patch to no longer use C-ism and removed the additional variants: if someone needs them at some point, I guess they'll add them. The true/false tests should likely ignore case though.

kduske commented on common/src/IO/EntParser.cpp:
```diff
@@ -622,6 +622,36 @@ std::optional<Color> EntParser::parseColor(
+  {
+    return true;
+  }
+  else if (strValue == "False")
```
You don't need the `else` here due to the unconditional return from the previous if block.

bmorel replied:
fun fact, that's the kind of things that usually makes me complain about dumb code, when I see a `else` after a `if(){return;}` :D

kduske commented on common/src/IO/EntParser.cpp:
```diff
@@ -622,6 +622,36 @@ std::optional<Color> EntParser::parseColor(
+  {
+    return false;
+  }
+  else
```
You can drop the else and the braces again because of the unconditional returns.

kduske reviewed (changes requested):
Thanks for addressing my feedback! I have a few more comments, though... sorry :-/

kduske commented:
@bmorel did you see my comments?

bmorel commented on common/src/IO/EntParser.cpp:
```diff
@@ -622,6 +622,36 @@ std::optional<Color> EntParser::parseColor(
+  const std::string& attributeName,
+  ParserStatus& /* status */)
+{
+  std::string strValue = element.Attribute(attributeName.c_str());
```
I will, since it is not my project, but if I may express my opinion, I think using `auto` all the time just ends up making the code harder to understand. I still use it, though, when the type can easily be guessed or is not important for the following operations, i.e. in range-for or with casts.
This is just my opinion, of course, and again: not my project, so I'll conform to codestyle.

bmorel commented:
No, I didn't connected to github, since github requires stupid 2FA to log. Same about my personal mails, for the same reason + gmail now refuses to obey the mail standards.
I'll fix those comments.

bmorel commented on common/src/IO/EntParser.cpp:
```diff
@@ -622,6 +622,36 @@ std::optional<Color> EntParser::parseColor(
+    return std::nullopt;
+  }
+
+  if (strValue == "True")
```
> I know I'm annoying

Not really. It's good to keep things consistent, makes projects easier to maintain, and even to mass-refactor when it's needed/useful.

Guess I need to use kdl::ci::str_mismatch then?

kduske replied:
No, `kdl::ci::str_is_equal` would be the function to use. That would make it completely case insensitive, which is also fine by me.

bmorel replied:
I think it's better to have complete insensitiveness than a partial one which will surprise people for no good reason.
I could not find str_is_equal in the kdl file I've looked... guess I missed it. Would have been easier to use indeed.

bmorel commented:
Addressed (hopefully, as CI does not says what would be expected!) the clang-format problem, I should take some time to make vim correctly detect the use of tab/spaces for indent someday.

bmorel commented:
CI (clang-format) still fails, but it is related to something I have *not* changed, so I could use some explanation here (and/or the fix directly, since CI is so long to complete... even the clang-format stuff)

kduske commented:
Hey, I'll just create a separate PR for this, I see that there is a lot of cruft in `EntParser.cpp` anyway and it needs to be cleaned up, so I'll do that and then add your new function on top of that PR. Thank you!

bmorel commented:
I would still be glad to know what does not passes here, if you have an idea?

kduske commented:
> > clang-format is complaining about EntParser.cpp, so I'm not sure what you mean.
> 
> It is complaining about the line where EntParser.h is included, to be exact, and since I didn't touched it, I don't understand what's the problem. Do you know how I could just summon clang-format on this file to get a diff, perhaps?

The idea is that you install clang-format and set up your IDE / editor to run it automatically when you save, that way you never have to worry about formatting again. Just make sure that you use the same version of clang-format that TB expects, which is 13.0. You can also find binaries in our [dev-tools repo](https://github.com/TrenchBroom/dev-tools).
